KBS: more prometheus metrics

pmores opened the pull request:
Add Prometheus metrics for authentication, attestation and policy evaluation as per discussion at issue #543.

The endpoint output after some successful resource accesses will be similar to
```
$ curl http://127.0.0.1:8080/metrics
# HELP attestation_errors_total Total count of errors during attestation processing
# TYPE attestation_errors_total counter
attestation_errors_total 0
# HELP attestation_requests_total Total count of attestation requests
# TYPE attestation_requests_total counter
attestation_requests_total 3
# HELP attestation_successes_total Total count of attestation successes
# TYPE attestation_successes_total counter
attestation_successes_total{tee_type="sample"} 3
# HELP auth_errors_total Total count of errors during auth processing
# TYPE auth_errors_total counter
auth_errors_total 0
# HELP auth_requests_total Total count of auth requests
# TYPE auth_requests_total counter
auth_requests_total 3
# HELP auth_successes_total Total count of successfully authenticated requests
# TYPE auth_successes_total counter
auth_successes_total 3
# HELP http_request_duration_seconds Distribution of request handling duration
# TYPE http_request_duration_seconds histogram
http_request_duration_seconds_bucket{le="0.0005"} 6
http_request_duration_seconds_bucket{le="0.001"} 7
http_request_duration_seconds_bucket{le="0.005"} 13
http_request_duration_seconds_bucket{le="0.01"} 13
http_request_duration_seconds_bucket{le="0.05"} 13
http_request_duration_seconds_bucket{le="0.5"} 13
http_request_duration_seconds_bucket{le="1"} 13
http_request_duration_seconds_bucket{le="+Inf"} 13
http_request_duration_seconds_sum 0.014657365
http_request_duration_seconds_count 13
# HELP http_request_size_bytes Distribution of request body sizes
# TYPE http_request_size_bytes histogram
http_request_size_bytes_bucket{le="32"} 8
http_request_size_bytes_bucket{le="128"} 17
http_request_size_bytes_bucket{le="512"} 23
http_request_size_bytes_bucket{le="2048"} 26
http_request_size_bytes_bucket{le="8192"} 26
http_request_size_bytes_bucket{le="+Inf"} 26
http_request_size_bytes_sum 7255
http_request_size_bytes_count 26
# HELP http_requests_total Total HTTP requests count
# TYPE http_requests_total counter
http_requests_total 13
# HELP http_response_size_bytes Distribution of response body sizes
# TYPE http_response_size_bytes histogram
http_response_size_bytes_bucket{le="32"} 0
http_response_size_bytes_bucket{le="128"} 0
http_response_size_bytes_bucket{le="512"} 0
http_response_size_bytes_bucket{le="2048"} 0
http_response_size_bytes_bucket{le="8192"} 0
http_response_size_bytes_bucket{le="+Inf"} 0
http_response_size_bytes_sum 0
http_response_size_bytes_count 0
# HELP kbs_policy_approvals_total Total count of requests approved by KBS policy
# TYPE kbs_policy_approvals_total counter
kbs_policy_approvals_total 3
# HELP kbs_policy_errors_total Total count of errors during KBS evaluation
# TYPE kbs_policy_errors_total counter
kbs_policy_errors_total 0
# HELP kbs_policy_evaluations_total Total count of KBS policy evaluations
# TYPE kbs_policy_evaluations_total counter
kbs_policy_evaluations_total 3
# HELP kbs_policy_violations_total Total count of requests denied by KBS policy
# TYPE kbs_policy_violations_total counter
kbs_policy_violations_total 0
# HELP resource_reads_total KBS resource read count
# TYPE resource_reads_total counter
resource_reads_total{resource_path="pvl/test/smazat"} 3
# HELP resource_writes_total KBS resource write count
# TYPE resource_writes_total counter
resource_writes_total{resource_path="pvl/test/smazat"} 1
```

Xynnn007 commented on kbs/src/attestation/backend.rs:
```diff
@@ -276,11 +279,13 @@ impl AttestationService {
-            .verify(tee, &nonce, &attestation_str)
-            .await
-            .context("verify TEE evidence failed")?;
+        let token = match self.inner.verify(tee, &nonce, &attestation_str).await {
+            Ok(token) => token,
+            Err(err) => {
+                ATTESTATION_FAILURES.inc();
+                return Err(err).context("verify TEE evidence failed");
+            }
+        };
```
There is a way for logging but I am not sure if the compiler will cry for this
```rust
self.inner
    .verify(tee, &nonce, &attestation_str)
    .await
    .inspect(|_| ATTESTATION_FAILURES.inc())
    .context("verify TEE evidence failed")?;
```

pmores replied:
That's a great point, the compiler doesn't mind it :-) and it actually works, too (provided I use `inspect_err()`, not `inspect()` which is only called if value is `Ok()`) - thanks!

Xynnn007 commented on kbs/src/attestation/backend.rs:
```diff
@@ -230,6 +231,8 @@ impl AttestationService {
         attestation: &[u8],
         request: HttpRequest,
     ) -> anyhow::Result<HttpResponse> {
+        ATTESTATION_REQUESTS.inc();
```
nit: Remote attestation (RCAR protocol) consists of two rounds of API calls, `auth` is the first round and `attest` is the second round. I'm not sure if this is appropriate here, maybe we can add some context like [here](https://github.com/confidential-containers/trustee/blob/main/kbs/src/plugins/implementations/resource/backend.rs#L129)?

pmores replied:
That makes it a lot clearer, thanks!  Wouldn't `auth` be best tracked then if we add a separate set of counters for it? If I understand correctly `auth` as such can't really fail (unless there is a request deserialisation or nonce generation error, or a version mismatch, it will always respond with a session cookie and thus succeed), `auth_requests_total` and `auth_requests_errors` whould seem enough.

Xynnn007 commented on kbs/src/policy_engine/opa/mod.rs:
```diff
@@ -33,6 +35,8 @@ impl PolicyEngineInterface for Opa {
         resource_path: &str,
         input_claims: &str,
     ) -> Result<bool, KbsPolicyEngineError> {
+        RESOURCE_POLICY_EVALS.inc();
```
We could move the two resource policy evals to a more [upper place](https://github.com/confidential-containers/trustee/blob/main/kbs/src/api_server.rs#L284-L286), as now policy engine is designed as a plugin interface. If a new plugin is introduced, it could help us save some code lines.

Xynnn007 reviewed (commented):
Thanks @pmores for this! some nits

Xynnn007 commented on kbs/src/api_server.rs:
```diff
@@ -280,8 +283,15 @@ pub(crate) async fn api(
+                    .await
+                    .inspect_err(|_| RESOURCE_POLICY_ERRORS.inc())?
+                {
+                    RESOURCE_POLICY_VIOLATIONS.inc();
```
The format of variable `endpoint` looks like `{plugin-name}/{other-parts}`. If you focus on resource related, you can count `endpoint` starts with `resource/` where `{plugin-name}` is `resource`. 

Btw, currently, this policy can indeed cover all plugin calls. The data input to the Policy Engine is a complete `endpoint`.  Thus KBS policy would be better? @fitzthum

pmores replied:
Okay, thanks for clearing this up!  I think I can just go ahead and do the rename since @fitzthum has already expressed his preference gently but firmly. :-)

Xynnn007 commented on kbs/src/prometheus/mod.rs:
```diff
@@ -57,6 +57,78 @@ lazy_static! {
+        Counter::with_opts(opts).unwrap()
+    };
+
+    /// KBS Auth Errors Total
+    pub(crate) static ref AUTH_ERRORS: Counter = {
+        let opts = Opts::new(
+            "auth_errors_total",
+            "Total count of errors during auth processing",
+        );
+        Counter::with_opts(opts).unwrap()
+    };
```
Oh, the same pattern of code breaks out here. It seems that we can consider using macros to simplify it. We can add it in the subsequent PR

Xynnn007 reviewed (commented):
Some comments. btw `auth` is good to carry an overall `inspect_err` as you suggested imo.

pmores commented:
Pushed the final version as far as I'm concerned, please do your final review.  Changes:
- renamed resource policy to KBS policy as per [review request](https://github.com/confidential-containers/trustee/pull/816#discussion_r2135970620)
- added success counters as per [review request](https://github.com/confidential-containers/trustee/pull/816#pullrequestreview-2910485317)
- added `tee_type` labelling to attestation success and failure counters as per [result of the pertinent discussion](https://github.com/confidential-containers/trustee/issues/543#issuecomment-2975898460)

Personally I'm not entirely happy with how attestation counters are handled but after several tests and attempts I believe any improvement is only possible with a fairly substantial refactor of `__attest()`/`attest()` which would not be appropriate at this time in this PR.

Xynnn007 commented on kbs/src/attestation/backend.rs:
```diff
@@ -268,31 +290,52 @@ impl AttestationService {
+            .inner
+            .verify(tee, &nonce, &attestation_str)
+            .await
+            .inspect(|_| {
+                ATTESTATION_SUCCESSES
+                    .with_label_values(&[&tee_type_label])
+                    .inc()
+            })
```
We can move this out of the nested `inspect` to keep the chain short.

pmores replied:
done

Xynnn007 commented on kbs/src/api_server.rs:
```diff
@@ -280,9 +283,18 @@ pub(crate) async fn api(
+                {
+                    KBS_POLICY_VIOLATIONS.inc();
                     return Err(Error::PolicyDeny);
+                } else {
+                    KBS_POLICY_APPROVALS.inc();
                 }
```
Some personal taste - we even do not need the `else` here

pmores replied:
done

fitzthum commented on kbs/src/attestation/backend.rs:
```diff
@@ -324,23 +349,41 @@ impl AttestationService {
+            .await
+            .inspect_err(|_| {
+                ATTESTATION_FAILURES
+                    .with_label_values(&[&tee_type_label])
```
At this point we aren't guaranteed to have a valid tee type so there is a chance of dos.

fitzthum replied:
actually maybe it's fine since we've deserialized the Tee enum

bpradipt commented:
I think having a readme for the metrics feature will be a good addition.

pmores commented:
please take a look

fitzthum commented:
We should add some info to the website about this.
